File: Chapter_2/2_1_4_add_binary_integers/sol1.py
'''
Same as leetcode-
https://leetcode.com/problems/add-binary

Runtime: 40 ms, faster than 72.14% of Python3 online submissions for Add Binary.
Memory Usage: 13.7 MB, less than 5.41% of Python3 online submissions for Add Binary.

'''

class Solution:
    def addBinary(self, a: str, b: str) -> str:
        i=len(a)-1
        j=len(b)-1
        carry=0
        result=''
        while (i>=0 or j>=0):
            sum=carry
            if (j>=0):
                sum=sum+int(b[j])
                j=j-1
            if (i>=0):
                sum=sum+int(a[i])
                i=i-1
            result=str(int(sum%2))+result
            carry=int(sum/2)
        if carry!=0:
            result=str(carry)+result
        return result



sol_obj=Solution()


a='1'
b='111'
a='101010101010'
b='101010101010'
a='101110'
b='110'
a='1110'
c=sol_obj.addBinary(a,b)
print (c)

File: Chapter_2/2_1_4_add_binary_integers/sol1.cpp
// Code to be added to this. Algo same as sol1.py. Refer
#include "string"
using namespace std;
class Solution
{
public:
    string addBinary(string a, string b)
    {
        string s = "";

        int carry = 0, i = a.size() - 1, j = b.size() - 1;
        int sum;
        while(i >= 0 || j >= 0)
        {
          sum=carry;
          if (i >= 0){
              sum =sum+a[i] - '0';
              i--;
          }
          if (j >= 0){
              sum =sum+b[j] - '0';
              j--;
          }

            s = char(sum % 2 + '0') + s;
            carry = sum/2;
        }
        if (carry == 1){
          s = char(carry % 2 + '0') + s;
        }

        return s;
    }
};

int main(int argc, char const *argv[]) {
  Solution obj;
  string a="100";
  string b="1";
  cout<<obj.addBinary(a,b)<<endl;
  return 0;
}
